File: src/game/room/room_model.c
#include "stdlib.h"
#include <ctype.h>

#include "room_model.h"
#include "shared.h"
#include "list.h"

#include "game/items/item.h"
#include "game/navigator/navigator_category_manager.h"

#include "game/items/items_data_parser.h"

/**
 *
 * @param modeltype
 * @param door_x
 * @param door_y
 * @param door_z
 * @param heightmap
 * @return
 */
room_model *room_model_create(char *model_id, char *model_name, int door_x, int door_y, double door_z, int door_dir, char *heightmap) {
    room_model *model = malloc(sizeof(room_model));
    model->model_id = strdup(model_id);
    model->model_name = strdup(model_name);
    model->door_x = door_x;
    model->door_y = door_y;
    model->door_z = door_z;
    model->door_dir = door_dir;
    model->map_size_x = 0;
    model->map_size_y = 0;
    model->heightmap = replace(heightmap, '|', "\r");
    
    List *items = item_parser_get_items(model->model_id);

    if (items != NULL) {
        model->public_items = items;
    } else {
        list_new(&model->public_items);
    }
    
    room_model_parse(model);
    return model;
}

void room_model_parse(room_model *room_model) {
    char *heightmap = strdup(room_model->heightmap);
    char *array[100];

    int lines = 0;

    array[lines] = strtok(heightmap,"\r");

    while(array[lines]!=NULL) {
        array[++lines] = strtok(NULL,"\r");
    }

    int map_size_x = strlen(array[0]);
    int map_size_y = lines;

    room_model->map_size_x = map_size_x;
    room_model->map_size_y = map_size_y;

    room_model->states = malloc(sizeof(*room_model->states) * map_size_x);
    room_model->heights = malloc(sizeof(*room_model->heights) * map_size_x);

    for (int x = 0; x < map_size_x ; x++) { 
         room_model->states[x] = malloc(sizeof(*room_model->states) * map_size_y);
         room_model->heights[x] = malloc(sizeof(*room_model->heights) * map_size_y);
    }


     for (int y = 0; y < map_size_y; y++) {
        char *line = array[y];

        for (int x = 0; x < strlen(line); x++) {
            char ch = (char)tolower(line[x]);

            if (ch == 'x') {
                room_model->states[x][y] = CLOSED;
                room_model->heights[x][y] = 0;
            } else {
                int height = ch - '0';
                room_model->states[x][y] = OPEN;
                room_model->heights[x][y] = height;
            }
            
            if (x == room_model->door_x && y == room_model->door_y) {
                room_model->states[x][y] = OPEN;
            }
        }
     }

    free(heightmap);
}

File: Kepler-Server/src/main/java/org/alexdev/kepler/game/room/models/RoomModel.java
package org.alexdev.kepler.game.room.models;

import org.alexdev.kepler.game.item.Item;
import org.alexdev.kepler.game.item.ItemPublicParser;
import org.alexdev.kepler.game.pathfinder.Position;
import org.alexdev.kepler.game.room.mapping.RoomTileState;
import org.alexdev.kepler.util.StringUtil;

import java.util.ArrayList;
import java.util.List;

public class RoomModel {
    private String modelId;
    private String modelName;
    private int doorX;
    private int doorY;
    private double doorZ;
    private int doorRotation;
    private int mapSizeX;
    private int mapSizeY;
    private String heightmap;
    private List<Item> publicItems;

    private RoomTileState[][] tileStates;
    private double[][] tileHeights;

    public RoomModel(String modelId, String modelName, int doorX, int doorY, double doorZ, int doorRotation, String heightmap) {
        this.modelId = modelId;
        this.modelName = modelName;
        this.doorX = doorX;
        this.doorY = doorY;
        this.doorZ = doorZ;
        this.doorRotation = doorRotation;
        this.heightmap = heightmap.replace("|", "\r");
        this.publicItems = ItemPublicParser.getPublicItems(this.modelId);
    }

    /**
     * Parse heightmap, add invalid tiles and the tile heights used
     * for walking, stairs, etc.
     */
    public void parse() {
        String[] lines = this.heightmap.split("\r");

        this.mapSizeX = lines.length;
        this.mapSizeY = lines[0].length();

        this.tileStates = new RoomTileState[this.mapSizeX][this.mapSizeY];
        this.tileHeights = new double[this.mapSizeX][this.mapSizeY];

        StringBuilder temporaryHeightmap = new StringBuilder();

        for (int x = 0; x < this.mapSizeX; x++) {
            String line = lines[x];

            for (int y = 0; y < this.mapSizeY; y++) {
                String tile = Character.toString(line.charAt(y));

                if (StringUtil.isNumber(tile)) {
                    this.tileStates[x][y] = RoomTileState.OPEN;
                    this.tileHeights[x][y] = Double.parseDouble(tile);
                } else {
                    this.tileStates[x][y] = RoomTileState.CLOSED;
                    this.tileHeights[x][y] = 0;
                }

                if (x == this.doorX && y == this.doorY) {
                    this.tileStates[x][y] = RoomTileState.OPEN;
                    this.tileHeights[x][y] = this.doorZ;
                }

                temporaryHeightmap.append(tile);
            }

            temporaryHeightmap.append("\r");
        }

        this.heightmap = temporaryHeightmap.toString();

    }

    /**
     * Get the tile state by given coordinates. This
     * doesn't include room furniture.
     *
     * @param x the x coordinate
     * @param y the y coordinate
     * @return the room state
     */
    public RoomTileState getTileState(int x, int y) {
        if (x < 0 || y < 0) {
            return RoomTileState.CLOSED;
        }

        if (x >= this.mapSizeX || y >= this.mapSizeY) {
            return RoomTileState.CLOSED;
        }

        return tileStates[x][y];
    }

    /**
     * Get the tile height, this doesn't include
     * furniture heights.
     *
     * @param x the x coordinate
     * @param y the y coordinate
     * @return the room height
     */
    public double getTileHeight(int x, int y) {
        if (x < 0 || y < 0) {
            return 0;
        }

        if (x >= this.mapSizeX || y >= this.mapSizeY) {
            return 0;
        }

        return tileHeights[x][y];
    }


    public String getId() {
        return modelId;
    }

    public String getName() {
        return modelName;
    }

    public Position getDoorLocation() {
        return new Position(this.doorX, this.doorY, this.doorZ, this.doorRotation, this.doorRotation);
    }

    public int getMapSizeX() {
        return mapSizeX;
    }

    public int getMapSizeY() {
        return mapSizeY;
    }

    public String getHeightmap() {
        return heightmap;
    }

    public List<Item> getPublicItems() {
        return publicItems;
    }
}